File: rust/RayTracing/src/ray_tracer/hittable/primitives/sphere.rs
use crate::ray_tracer::hittable::{Hittable, HitRecord, material::Material};
use nalgebra_glm as glm;

use super::primitive::Primitive;

pub struct Sphere
{
    pub origin: glm::Vec3,
    pub radius: f32,
    pub mat: Box<Material>
}

impl Hittable for Sphere
{
    fn intersect<'a>(&'a self, ray: &crate::ray_tracer::ray::Ray, min: f32, max: f32) -> Option<HitRecord<'a>> 
    {
        let oc = ray.orig - self.origin;
        let a = ray.dir.magnitude_squared();
        let half_b = glm::dot(&oc, &ray.dir);
        let c = oc.magnitude_squared() - self.radius*self.radius;

        let discriminant = half_b*half_b - a*c;
        
        if discriminant < 0.0
        {
            return None;
        }
        let sqrtd = discriminant.sqrt();

        let mut root = (-half_b - sqrtd)/a;
        if root < min || max < root
        {
            root = (-half_b + sqrtd) / a;
            if root < min || max < root
            {
                return None;
            }
        }

        let pos = ray.get_at(root);
        let norm = (pos-self.origin) / self.radius;
        
        return Some(HitRecord::new(&pos, &norm, root, ray, self.mat.as_ref()));
    }
}

impl Into<Primitive> for Sphere
{
    fn into(self) -> Primitive {
        Primitive::Sphere(self)
    }
}

File: python/RayTracing/src/ray_tracer/primitive/sphere.py
from typing import Optional, Tuple
import numpy as np
import numba as nb
from ray_tracer.primitive.material import Material
from ray_tracer.primitive.base import Hittable, Hit
from ray_tracer.ray import RayWrapper

class Sphere(Hittable):
    @staticmethod
    @nb.njit
    def __sphere_intersection(self_pos: np.ndarray, self_radius: float, ray_dir:np.ndarray, ray_pos:np.ndarray, min: float, max: float) -> Optional[Tuple[np.ndarray, np.ndarray, float]]:
        oc = ray_pos - self_pos
        a = np.sum(ray_dir*ray_dir)
        half_b = np.sum(oc*ray_dir)
        c = np.sum(oc*oc) - self_radius**2
        
        discriminant = half_b*half_b - a*c

        if discriminant < 0.0:
            return None
        sqrtd = discriminant**0.5

        root = (-half_b - sqrtd) / a
        if root < min or max < root:
            root = (-half_b + sqrtd) / a
            if root < min or max < root:
                return None
        
        pos = ray_pos + root * ray_dir
        norm = (pos-self_pos) / self_radius

        return (pos, norm, root)
    def __init__(self, pos: np.ndarray, radius: float, mat: Material):
        self.pos = pos
        self.radius = radius
        self.mat = mat
    def intersect(self, ray, min, max) -> Optional[Hit]:
        ray = RayWrapper(ray)

        data = Sphere.__sphere_intersection(self.pos, self.radius, ray.dir, ray.pos, min, max)

        if data is None:
            return None
        else:
            pos, norm, root = data
            return Hit(pos, norm, root, ray, self.mat)
